File: scala-server/app/com/bolour/eagerwords/server/domain/Strip.scala
/*
 * Copyright 2017-2020 Azad Bolour
 * Licensed under GNU Affero General Public License v3.0 -
 *    https://github.com/azadbolour/eagerwords/blob/master/LICENSE.md
 */
package com.bolour.eagerwords.server.domain

import com.bolour.eagerwords.server.util.WordUtil
import com.bolour.plane.domain.Axis._
import com.bolour.plane.domain.Axis.Axis
import com.bolour.eagerwords.server.util.WordUtil._
import com.bolour.plane.domain.{Axis, Point}
import com.bolour.util.BlackWhite

/**
  * A strip of the board - a horizontal or vertical line segment.
  *
  * @param axis X or Y for horizontal or vertical.
  * @param lineNumber The index of the line on the board - for example 0 for the first row.
  * @param begin The offset within the given line where the strip begins.
  * @param end The offset within the given line where the strip ends (inclusive).
  * @param content The content of the strip - sequence of letters and blanks.
  */
case class Strip(
  axis: Axis,
  lineNumber: Int,
  begin: Int,
  end: Int,
  content: String
) {

  /**
    * Combination of letters on strip (sorted - with dups).
    */
  val letters: LetterCombo = stringToLetterCombo(content.filter(_ != blankChar))

  /**
    * Number of blank slots on strip.
    */
  val numBlanks: Int = content.length - letters.length

  val len: Int = end - begin + 1

  /**
    * Word can potentially be played to this strip.
    * They have the same length, and the non-blank slots
    * of the strip match the corresponding letter of the word.
    */
  def admits(word: String): Boolean = {
    (len == word.length) && fits(content, word)
  }

  def findFittingWords(words: List[DictWord]): List[DictWord] = {
    words.filter(admits)
  }

  def findFittingWord(words: List[DictWord]): Option[DictWord] =
    words.find(admits)

  def hasAnchor: Boolean = numBlanks < len

  def isDense(maxBlanks: Int): Boolean = hasAnchor && numBlanks <= maxBlanks

  def row(offset: Int): Int = axis match {
    case X => lineNumber
    case Y => begin + offset
  }

  def column(offset: Int): Int = axis match {
    case X => begin + offset
    case Y => lineNumber
  }

  def point(offset: Int) = Point(row(offset), column(offset))

  def offset(point: Point): Int = {
    val indexInLine = axis match {
      case Axis.X => point.col
      case Axis.Y => point.row
    }
    indexInLine - begin
  }

  def fillBlankInStrip(point: Point, ch: Char): String = content.updated(offset(point), ch)

  // TODO. Use fold.
  /**
    * It has already been established that the rest of the word has the same length
    * as the rest of the strip content - so just compare their corresponding letters
    */
  private def fits(restContent: String, restWord: String): Boolean =
    if (restWord.isEmpty) true
    else Strip.fitsSlot(restContent.head, restWord.head) &&
           fits(restContent.tail, restWord.tail)

  def blankPoints: List[Point] = {
    val blankOffsets = content.indices.toList.filter(offset => WordUtil.isBlankChar(content(offset)))
    blankOffsets map point
  }
}

object Strip {

  def lineStrip(axis: Axis, lineNumber: Int, line: String, begin: Int, end: Int): Strip = {
    val content = line.slice(begin, end + 1)
    Strip(axis, lineNumber, begin, end, content)
  }

  def fitsSlot(slotLetter: Char, wordLetter: Char): Boolean =
    slotLetter == ' ' || slotLetter == wordLetter

  def stripFromBlackWhiteLine(axis: Axis, lineNumber: Int, blackWhiteChars: List[BlackWhite[Char]], offset: Int, size: Int): Strip = {
    val lineAsString = (blackWhiteChars map blackWhiteToChar).mkString
    lineStrip(axis, lineNumber, lineAsString, offset, offset + size - 1)
  }

  def stripsInLine(axis: Axis, lineNumber: Int, chars: String): List[Strip] = {
    val dimension = chars.length
    for {
      offset <- (0 until dimension).toList
      end <- offset until dimension
    } yield lineStrip(axis, lineNumber, chars, offset, end)
  }

  def allStripsInBlackWhiteLine(axis: Axis, lineNumber: Int, line: List[BlackWhite[Char]]): List[Strip] = {
    val lineAsString = (line map blackWhiteToChar).mkString
    stripsInLine(axis, lineNumber, lineAsString)
  }

  def blackWhiteToChar(blackWhiteChar: BlackWhite[Char]): Char =
    blackWhiteChar.toValueWithDefaults(blackChar, blankChar)
}

File: haskell-server/src/EagerWords/Server/Domain/Strip.hs
--
-- Copyright 2017-2020 Azad Bolour
-- Licensed under GNU Affero General Public License v3.0 -
--   https://github.com/azadbolour/eagerwords/blob/master/LICENSE.md
--

{-# LANGUAGE NamedFieldPuns #-}
{-# LANGUAGE DisambiguateRecordFields #-}
{-# LANGUAGE RecordWildCards #-}

module EagerWords.Server.Domain.Strip (
    Strip(..)
  , GroupedStrips
  , mkStrip
  , stripPoint
  , stripLength
  , lineStrip
  , charLineSegmentToStrip
  , stripFromBlackWhiteLine
  , allStripsInBlackWhiteLine
  , hasAnchor
  , hasBlanks
  , pointAtOffset
  , stripsInLine
  , blankPoints
  , isDense
  , offset
  , fillBlankInStrip
  ) where

import qualified Data.List as List
import Data.ByteString.Char8 (ByteString)
import Data.Map (Map)
import qualified Data.Map as Map

import qualified Bolour.Plane.Domain.LineSegment as LineSegment
import Bolour.Plane.Domain.LineSegment (LineSegment, LineSegment(LineSegment))
import Bolour.Util.BlackWhite (BlackWhite, BlackWhite(Black, White))
import qualified Bolour.Util.BlackWhite as BlackWhite
import Bolour.Language.Util.WordUtil (DictWord, LetterCombo, BlankCount, ByteCount)
import qualified Bolour.Language.Util.WordUtil as WordUtil
import qualified Bolour.Util.MiscUtil as MiscUtil
import Bolour.Plane.Domain.Axis (Axis, Coordinate)
import Bolour.Plane.Domain.Point (Point, Point(Point))
import qualified Bolour.Plane.Domain.Point as Point
import qualified Bolour.Plane.Domain.Axis as Axis

-- | A horizontal or vertical strip of the board.
data Strip = Strip {
    axis :: Axis              -- ^ direction of the strip - X = horizontal, Y = vertical.
  , lineNumber :: Coordinate  -- ^ position of the strip's line within all lines in teh same direction
  , begin :: Coordinate       -- ^ beginning index of the strip
  , end :: Coordinate         -- ^ ending index of the strip
  , content :: String         -- ^ letters and blanks
  , letters :: LetterCombo    -- ^ existing combination of letters in the strip
  , blanks :: BlankCount      -- ^ number of blank spaces in the strip
} deriving (Eq, Show)

mkStrip :: Axis -> Int -> Int -> Int -> String -> Strip
mkStrip axis lineNumber begin end content =
  Strip axis lineNumber begin end content (nonBlankCombo content) (numBlanks content)

-- | Strips of a a board grouped by length and by number of blanks.
--   They are grouped by length to allow longer strips to be matched first (they are of highest value).
--   They are grouped by the number of blanks so that a given combination of tray letters can be tested
--   only against those strips that have the right number of blanks (= the size of the combination).
type GroupedStrips = Map ByteCount (Map BlankCount [Strip])


-- | Does a given word fit exactly in a strip.
--   The word have have been chosen to have the combination of the strip's letters
--   and a combination of tray letters equal in size to the strip's blanks.
--   The word would have been some permutation of the combined strip and tray letters.
--   That permutation may or may not match the strips existing letters.
matchWordToStrip :: Strip -> DictWord -> Maybe (Strip, DictWord)

matchWordToStrip strip word = Nothing

-- | Get one matching word to the strip among the given words.
--   All the words being tested would have had the same length.
--   So they would all have the same score and any one will do.
matchWordsToStrip :: Strip -> [DictWord] -> Maybe (Strip, DictWord)

matchWordsToStrip strip words = Nothing

numBlanks :: String -> Int
numBlanks string = length $ filter WordUtil.isBlankChar string

nonBlankCombo :: String -> LetterCombo
nonBlankCombo string =
  let nonBlanks = filter (not . WordUtil.isBlankChar) string
  in WordUtil.mkLetterCombo nonBlanks

charLineSegmentToStrip :: LineSegment Char -> Strip
charLineSegmentToStrip LineSegment {axis, lineNumber, begin, end, segment} =
  let content = maybeCharToChar <$> segment
  in mkStrip axis lineNumber begin end content

lineStrip :: Axis -> Coordinate -> String -> Int -> ByteCount -> Strip
lineStrip axis lineNumber line offset size =
  mkStrip axis lineNumber offset (offset + size - 1) stringContent
    where stringContent = (take size . drop offset) line

blackWhiteToChar :: BlackWhite Char -> Char
blackWhiteToChar blackWhiteChar = BlackWhite.toValueWithDefaults blackWhiteChar WordUtil.blackChar WordUtil.blankChar

maybeCharToChar :: Maybe Char -> Char
maybeCharToChar Nothing = WordUtil.blankChar
maybeCharToChar (Just ch) = ch

stripFromBlackWhiteLine :: Axis -> Coordinate -> [BlackWhite Char] -> Int -> ByteCount -> Strip
stripFromBlackWhiteLine axis lineNumber blackWhites offset size =
  let lineAsString = blackWhiteToChar <$> blackWhites
  in lineStrip axis lineNumber lineAsString offset size

allStripsInBlackWhiteLine :: Axis -> Coordinate -> [BlackWhite Char] -> [Strip]
allStripsInBlackWhiteLine axis lineNumber blackWhites =
  let lineAsString = blackWhiteToChar <$> blackWhites
  in stripsInLine axis lineNumber lineAsString

stripsInLine :: Axis -> Int -> String -> [Strip]
stripsInLine axis lineNumber chars = do
  let dimension = length chars
  offset <- [0 .. (dimension - 1)]
  size <- [1 .. (dimension - offset - 1)]
  return $ lineStrip axis lineNumber chars offset size

stripPoint :: Strip -> Coordinate -> Point
stripPoint Strip {axis, lineNumber, begin} offset =
  case axis of
  Axis.X -> Point lineNumber (begin + offset)
  Axis.Y -> Point (begin + offset) lineNumber

hasAnchor :: Strip -> Bool
hasAnchor strip @ Strip { letters } = length letters > 0

hasBlanks :: Strip -> Bool
hasBlanks Strip {blanks} = blanks > 0

isDense :: Strip -> Int -> Bool
isDense strip @ Strip {blanks} maxBlanks = hasAnchor strip && blanks <= maxBlanks

blankPoints :: Strip -> [Point]
blankPoints strip @ Strip {content} =
  let blankOffsets = filter (\offset -> WordUtil.isBlankChar (content !! offset)) [0 .. length content - 1]
  in stripPoint strip <$> blankOffsets

-- TODO. Redundant. See stripPoint.
pointAtOffset :: Strip -> Int -> Point
pointAtOffset Strip {lineNumber, begin, axis} offset =
  case axis of
    Axis.X -> Point lineNumber (begin + offset)
    Axis.Y -> Point (begin + offset) lineNumber

stripLength :: Strip -> Int
stripLength Strip {begin, end} = end - begin + 1

offset :: Strip -> Point -> Int
offset Strip {begin, axis} Point {row, col} =
    let indexInLine = case axis of
                      Axis.X -> col
                      Axis.Y -> row
    in indexInLine - begin

fillBlankInStrip :: Point -> Char -> Strip -> String
fillBlankInStrip point ch strip @ Strip { content }=
  MiscUtil.setListElement content (offset strip point) ch

